### PR TITLE
Add normalizePath and tests

### PR DESCRIPTION
**Pull request by FedericoCeratto**
Related to #6486

### CODE REVIEW

**Comment by yglukhov** [comment]
Thats cool, but maybe its better to move `normalizePath` to `ospaths.nim`? Also please have a look at  https://github.com/yglukhov/nimx/blob/master/nimx/pathutils.nim#L33. That is an inplace version to avoid allocations. Immutable version might be called `normalizedPath` and reuse the inplace version. Also my impl may optionally respect native path separators which proved to be quite useful.

**Comment by jxy** [comment]
Any news on this?  I would like these to be in the stdlib somewhere.

**Comment by kaushalmodi** [comment]
Just recently, I wished such a function existed that converted relative paths like `../foo/` to absolute paths. 

Can this please be added?

**Comment by dom96** [comment]
This is great, but it sucks that ``os`` in general doesn't offer in-place versions of procs. Providing it for this is a bit awkward.

**Comment by FedericoCeratto** [comment]
Rebased. Feel free to merge it as it is or modify the PR.

**Comment by dom96** [comment]
Thanks for rebasing, but tests are failing.

**Comment by dom96** [comment]
The ``tos`` test is failing so your new tests or implementation must be wrong.

**Comment by timotheecour** [comment]
just sent out https://github.com/nim-lang/Nim/pull/8213 add os.absolutePath; fixes #8174
it allows passing in a 2nd (optional) argument (root), and works correctly with cases where `path` is already absolute, ignoring the `root` argument in that case

**Comment by Araq** [comment]
Is ``absolutePath`` the same as  ``result = if isAbsolute(x): x else: getCurrentDir() / x`` ?

File: lib/pure/os.nim [hunk 1]
```diff
@@ -298,8 +298,11 @@ proc setCurrentDir*(newDir: string) {.inline, tags: [].} =
 
 proc expandFilename*(filename: string): string {.rtl, extern: "nos$1",
   tags: [ReadDirEffect].} =
-  ## Returns the full (`absolute`:idx:) path of the file `filename`,
-  ## raises OSError in case of an error.
+  ## Returns the full (`absolute`:idx:) path of an existing file `filename`,
+  ## raises OSError in case of an error. Follows symlinks.
+  ##
+  ## To create absolute paths from any path (existing or not) see
+  ## `<#absolutePath>`_.
```
Comment (dom96) [inline lib/pure/os.nim]: I think ``expandPath`` would make more sense as a name here. ``absolutePath`` makes me wonder how it differs from ``normalizePath``.

File: lib/pure/os.nim [hunk 2]
```diff
@@ -338,6 +341,62 @@ proc expandFilename*(filename: string): string {.rtl, extern: "nos$1",
...
+  ## Normalize a path.
+  ##
+  ## Consecutive directory separators are collapsed, including an initial double slash.
+  ##
+  ## On relative paths, double dot (..) sequences are collapsed if possible.
+  ## On absolute paths they are always collapsed.
+  ##
+  ## Warning: URL-encoded and Unicode attempts at directory traversal are not detected.
+  ## Triple dot is not handled.
+  let is_abs = path[0] == separator
```
Comment (dom96) [inline lib/pure/os.nim]: not a big deal, but convention in stdlib is `camelCase`.

Comment (dom96) [reply lib/pure/os.nim]: Also, on Windows this is wrong, no? You should use the ``isAbsolute`` procedure.

File: tests/stdlib/tos.nim [hunk 3]
```diff
@@ -138,4 +139,53 @@ let tm = fromUnix(0) + 100.microseconds
...
 echo getLastModificationTime("a") == tm
-removeFile("a")
\ No newline at end of file
+removeFile("a")
+
+block normalizedPath:
+  doAssert normalizedPath("/a/b/../../../foo", '/') == "/foo"
+  doAssert normalizedPath("\\a\\b\\..\\..\\..\\foo", '\\') == "\\foo"
+
+when defined(Linux) or defined(osx):
```
Comment (dom96) [inline tests/stdlib/tos.nim]: That's cheating. You can't add something to ``os`` and only test it on Linux/Mac.

File: lib/pure/os.nim [hunk 4]
```diff
@@ -338,6 +341,62 @@ proc expandFilename*(filename: string): string {.rtl, extern: "nos$1",
...
+      stack.add(p)
+
+  if is_abs:
+    path = separator & join(stack, $separator)
+  elif stack.len > 0:
+    path = join(stack, $separator)
+  else:
+    path = "."
+
+proc normalizedPath*(path: string, separator=DirSep): string {.rtl, extern: "nos$1", tags: [].} =
```
Comment (dom96) [inline lib/pure/os.nim]: Can you remember where in the stdlib currently this convention is in use? It's about time this is documented but I can't find any other module that uses it.

File: tests/stdlib/tos.nim [hunk 5]
```diff
@@ -138,4 +139,53 @@ let tm = fromUnix(0) + 100.microseconds
...
+      doAssert normalizedPath("./") == "."
+
+    block absolute:
+      doAssert normalizedPath("/") == "/"
+      doAssert normalizedPath("/.") == "/"
+      doAssert normalizedPath("/..") == "/"
+      doAssert normalizedPath("/../") == "/"
+      doAssert normalizedPath("/../..") == "/"
+      doAssert normalizedPath("/../../") == "/"
+      doAssert normalizedPath("/../../../") == "/"
```
Comment (dom96) [inline tests/stdlib/tos.nim]: I wonder if it would make more sense to raise an error for these.

**Comment by dom96** [comment]
I do also prefer the ``absolutePath`` implementation that @timotheecour made in #8213.

**Comment by FedericoCeratto** [comment]
@Araq absolutePath is meant to do that (see implementation) and it's also normalizing the output by trimming away a trailing separator.

**Comment by FedericoCeratto** [comment]
@dom96 "expansion" usually refers to expanding "~" into the home directory and so on.
absolutePath is meant to go from relative to absolute path, hence the name.

**Comment by kaushalmodi** [comment]
@dom96 I agree on the "absolutePath" naming. I don't see any confusion when one sees this name. I infact ended up on this issue thread when I couldn't find anything named "absolutePath" in the stdlib.

**Comment by FedericoCeratto** [comment]
See expandFilename() discussion in #6486 . Unlike the proposed absolutePath, expandFilename() is acting on existing files and directories only.

File: tests/stdlib/tos.nim (hunk 5 above)
Comment (FedericoCeratto) [reply tests/stdlib/tos.nim]: That's the point of normalization. If you are looking for a way to detect and prevent directory traversal I think that should go into joinPath

**Comment by dom96** [comment]
Okay, ``absolutePath`` it is. Can anyone point me to where this `normalize` vs. `normalized` convention is used? I can't find it in the stdlib and I want to document it.

File: lib/pure/os.nim [hunk 6]
```diff
@@ -338,6 +338,47 @@ proc expandFilename*(filename: string): string {.rtl, extern: "nos$1",
       result = $r
       c_free(cast[pointer](r))
 
+proc normalizePath*(path: var string) {.rtl, extern: "nos$1", tags: [].} =
```
Comment (timotheecour) [inline lib/pure/os.nim]: this should be in ospaths, no? nothing in `normalizePath` accesses file system

Comment (dom96) [reply lib/pure/os.nim]: oh yeah, I forgot about this. But @FedericoCeratto has waited long enough with this PR so we can fix this in a separate PR.

Comment (timotheecour) [reply lib/pure/os.nim]: of course; I much prefer faster PR turnaround than waiting forever for PR's to be so-called "perfect" (this slow turnaround really hurts D ecosystem IMO, Nim seems better in that regard)

Comment (timotheecour) [reply lib/pure/os.nim]: => kept track of that in https://github.com/nim-lang/Nim/issues/8177 meta-issue so we don't forget

**Pull request merged**